src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/CallConverterThunk.cs

```diff
@@ -1128,15 +1129,36 @@ unsafe private static void InvokeTarget(void* allocatedStackBuffer, ref CallConv
                         if (!conversionParams._callerArgs.HasRetBuffArg() && conversionParams._calleeArgs.HasRetBuffArg())
                             returnValueToCopy = (void*)(new IntPtr(*((void**)returnValueToCopy)) + IntPtr.Size);
 
+                        if (returnType == CorElementType.ELEMENT_TYPE_BYREF)
+                        {
+                            // If this is a byref return, we're going to dereference the result
+                            returnValueToCopy = *(void**)returnValueToCopy;
+                        }
+
+                        RuntimeTypeHandle returnTypeRuntimeTypeHandle = thRetType.GetRuntimeTypeHandle();
+
                         // Need to box value type before returning it
                         object returnValue;
-                        if (thRetType.IsPointerType())
+                        if (returnType == CorElementType.ELEMENT_TYPE_BYREF && returnValueToCopy == null)
+                        {
+                            // This is a byref return and dereferencing it would result in a NullReferenceException.
+                            // Set the return value to a sentinel that InvokeUtils will recognize.
+                            // Can't throw from here or we would wrap this in a TargetInvocationException.
+                            returnValue = InvokeUtils.NullByRefValueSentinel;
+                        }
+                        else if (RuntimeAugments.IsUnmanagedPointerType(returnTypeRuntimeTypeHandle))
+                        {
+                            returnValue = System.Reflection.Pointer.Box(*(void**)returnValueToCopy, Type.GetTypeFromHandle(returnTypeRuntimeTypeHandle));
+                        }
+                        else if (RuntimeAugments.IsValueType(returnTypeRuntimeTypeHandle))
                         {
-                            returnValue = System.Reflection.Pointer.Box(*(void**)returnValueToCopy, Type.GetTypeFromHandle(thRetType.GetRuntimeTypeHandle()));
+                            returnValue = RuntimeAugments.Box(returnTypeRuntimeTypeHandle, new IntPtr(returnValueToCopy));
                         }
                         else
                         {
-                            returnValue = RuntimeAugments.Box(thRetType.GetRuntimeTypeHandle(), new IntPtr(returnValueToCopy));
+                            // byref return of a reference type
```

Comment:
> reference type [](start = 48, length = 15)

I think byref valuetypes need to be handled differently. Based on your comment above "if this is a byref return, we're going to dereference the result", and the fact that value types need to be boxed before returning them, shouldn't we box here too?

Reply:
We'll box them in the `if (RuntimeAugments.IsValueType(returnTypeRuntimeTypeHandle))` branch a couple lines above. The type handle used for the check is the type handle of the return type, but without the byrefness.

This branch really only handles the reference type case. Thinking about it, I can probably just make this call `RhBoxAny` and get rid of the `IsValueType` check because that one handles both reference types and valuetypes.

Reply:
You can use the RhBoxAny if you wish. Just add a comment in this file to clarify that RhBoxAny will do the IsValueType check, to make this piece of code easier to read :)